From 173e4007472ae094222ff7d74e3b1c8933027b8a Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Thu, 27 Nov 2014 04:03:15 +0200
Subject: [PATCH 584/827] v4l: xilinx: rgb2yuv: Use xvip_init_resources()

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Replace the manual ioremap by a call to xvip_init_resources(), which
also handles the functional clock.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
(cherry picked from commit 1b82e544d242a8d372b1fc36470d46385611cdc0)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 .../devicetree/bindings/media/xilinx/xlnx,v-rgb2yuv.txt     |  3 +++
 drivers/media/platform/xilinx/xilinx-rgb2yuv.c              | 13 +++++++------
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-rgb2yuv.txt b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-rgb2yuv.txt
index 0c843bc..d28b495 100644
--- a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-rgb2yuv.txt
+++ b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-rgb2yuv.txt
@@ -7,6 +7,8 @@ Required properties:
 
 - reg: Physical base address and length of the registers set for the device.
 
+- clocks: Reference to the video core clock.
+
 - ports: Video ports, using the DT bindings defined in ../video-interfaces.txt.
   The rgb2yuv has an input port (0) and an output port (1).
 
@@ -22,6 +24,7 @@ Example:
 	axi_rgb2yuv_0: axi_rgb2yuv {
 		compatible = "xlnx,v-rgb2yuv-7.1";
 		reg = <0x40100000 0x10000>;
+		clocks = <&clkc 15>;
 
 		ports {
 			#address-cells = <1>;
diff --git a/drivers/media/platform/xilinx/xilinx-rgb2yuv.c b/drivers/media/platform/xilinx/xilinx-rgb2yuv.c
index e10cd89..0a6662a 100644
--- a/drivers/media/platform/xilinx/xilinx-rgb2yuv.c
+++ b/drivers/media/platform/xilinx/xilinx-rgb2yuv.c
@@ -437,7 +437,6 @@ static int xrgb2yuv_parse_of(struct xrgb2yuv_device *xrgb2yuv)
 static int xrgb2yuv_probe(struct platform_device *pdev)
 {
 	struct xrgb2yuv_device *xrgb2yuv;
-	struct resource *res;
 	struct v4l2_subdev *subdev;
 	struct v4l2_mbus_framefmt *default_format;
 	unsigned int i;
@@ -453,10 +452,9 @@ static int xrgb2yuv_probe(struct platform_device *pdev)
 	if (ret < 0)
 		return ret;
 
-	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
-	xrgb2yuv->xvip.iomem = devm_ioremap_resource(&pdev->dev, res);
-	if (IS_ERR(xrgb2yuv->xvip.iomem))
-		return PTR_ERR(xrgb2yuv->xvip.iomem);
+	ret = xvip_init_resources(&xrgb2yuv->xvip);
+	if (ret < 0)
+		return ret;
 
 	/* Reset and initialize the core */
 	xvip_reset(&xrgb2yuv->xvip);
@@ -490,7 +488,7 @@ static int xrgb2yuv_probe(struct platform_device *pdev)
 	subdev->entity.ops = &xrgb2yuv_media_ops;
 	ret = media_entity_init(&subdev->entity, 2, xrgb2yuv->pads, 0);
 	if (ret < 0)
-		return ret;
+		goto error;
 
 	v4l2_ctrl_handler_init(&xrgb2yuv->ctrl_handler, 13);
 
@@ -523,6 +521,7 @@ static int xrgb2yuv_probe(struct platform_device *pdev)
 error:
 	v4l2_ctrl_handler_free(&xrgb2yuv->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
+	xvip_cleanup_resources(&xrgb2yuv->xvip);
 	return ret;
 }
 
@@ -535,6 +534,8 @@ static int xrgb2yuv_remove(struct platform_device *pdev)
 	v4l2_ctrl_handler_free(&xrgb2yuv->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
 
+	xvip_cleanup_resources(&xrgb2yuv->xvip);
+
 	return 0;
 }
 
-- 
2.9.3

